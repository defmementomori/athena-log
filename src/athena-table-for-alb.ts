import { aws_glue as glue, Stack } from 'aws-cdk-lib';
import { Construct } from 'constructs';

/**
 * Properties for the AthenaTableForAlb construct.
 */
export interface AthenaTableForAlbProps {
  /**
   * The name of the S3 bucket where the ALB access logs are stored.
   */
  readonly logBucketName: string;

  /**
   * The name of the Glue database where the table will be created.
   * The user is responsible for ensuring this database exists.
   */
  readonly databaseName: string;

  /**
   * The name for the Glue table.
   */
  readonly tableName: string;

  /**
   * The start date for the partition projection.
   * You must provide a value in 'YYYY/MM/DD' format.
   */
  readonly projectionStartDate: string;

  /**
   * The prefix where ALB access logs are stored within the bucket.
   * e.g., if logs are at `s3://my-bucket/my-prefix/AWSLogs/...`, specify 'my-prefix'.
   * @default Logs are expected at the root level (`s3://bucket-name/AWSLogs/...`).
   */
  readonly logPrefix?: string;
}

/**
 * A CDK construct to create an AWS Glue table for querying ALB access logs with Athena.
 * It automatically configures partition projection by date, simplifying log analysis.
 */
export class AthenaTableForAlb extends Construct {
  /**
   * The created Glue Table resource.
   */
  public readonly table: glue.CfnTable;

  constructor(scope: Construct, id: string, props: AthenaTableForAlbProps) {
    super(scope, id);

    const { logBucketName, databaseName, tableName, projectionStartDate, logPrefix } = props;

    // Add a runtime check for all required properties for extra safety.
    if (!logBucketName || !databaseName || !tableName || !projectionStartDate) {
      throw new Error('The `logBucketName`, `databaseName`, `tableName`, and `projectionStartDate` properties are all required.');
    }

    const stack = Stack.of(this);

    // Build the S3 path to the logs dynamically
    const s3LocationPrefix = logPrefix ? `${logPrefix}/` : '';
    const s3Location = `s3://${logBucketName}/${s3LocationPrefix}AWSLogs/${stack.account}/elasticloadbalancing/${stack.region}/`;

    this.table = new glue.CfnTable(this, 'Default', {
      catalogId: stack.account,
      databaseName: databaseName,
      tableInput: {
        name: tableName,
        description: 'Table for querying ALB access logs',
        tableType: 'EXTERNAL_TABLE',
        parameters: {
          'projection.enabled': 'true',
          'projection.day.type': 'date',
          'projection.day.range': `${projectionStartDate},2200`,
          'projection.day.format': 'yyyy/MM/dd',
          'projection.day.interval': '1',
          'projection.day.interval.unit': 'DAYS',
          'storage.location.template': `${s3Location}\${day}/`,
          'EXTERNAL': 'TRUE',
          'has_encrypted_data': 'false',
        },
        partitionKeys: [{ name: 'day', type: 'string' }],
        storageDescriptor: {
          columns: [
            { name: 'type', type: 'string' },
            { name: 'time', type: 'string' },
            { name: 'elb', type: 'string' },
            { name: 'client_ip', type: 'string' },
            { name: 'client_port', type: 'int' },
            { name: 'target_ip', type: 'string' },
            { name: 'target_port', type: 'int' },
            { name: 'request_processing_time', type: 'double' },
            { name: 'target_processing_time', type: 'double' },
            { name: 'response_processing_time', type: 'double' },
            { name: 'elb_status_code', type: 'int' },
            { name: 'target_status_code', type: 'string' },
            { name: 'received_bytes', type: 'bigint' },
            { name: 'sent_bytes', type: 'bigint' },
            { name: 'request_verb', type: 'string' },
            { name: 'request_url', type: 'string' },
            { name: 'request_proto', type: 'string' },
            { name: 'user_agent', type: 'string' },
            { name: 'ssl_cipher', type: 'string' },
            { name: 'ssl_protocol', type: 'string' },
            { name: 'target_group_arn', type: 'string' },
            { name: 'trace_id', type: 'string' },
            { name: 'domain_name', type: 'string' },
            { name: 'chosen_cert_arn', type: 'string' },
            { name: 'matched_rule_priority', type: 'string' },
            { name: 'request_creation_time', type: 'string' },
            { name: 'actions_executed', type: 'string' },
            { name: 'redirect_url', type: 'string' },
            { name: 'lambda_error_reason', type: 'string' },
            { name: 'target_port_list', type: 'string' },
            { name: 'target_status_code_list', type: 'string' },
            { name: 'classification', type: 'string' },
            { name: 'classification_reason', type: 'string' },
            { name: 'conn_trace_id', type: 'string' },
          ],
          location: s3Location,
          inputFormat: 'org.apache.hadoop.mapred.TextInputFormat',
          outputFormat: 'org.apache.hadoop.hive.ql.io.HiveIgnoreKeyTextOutputFormat',
          serdeInfo: {
            serializationLibrary: 'org.apache.hadoop.hive.serde2.RegexSerDe',
            parameters: {
              'serialization.format': '1',
              'input.regex':
                '([^ ]*) ([^ ]*) ([^ ]*) ([^ ]*):([0-9]*) ([^ ]*)[:-]([0-9]*) ([-.0-9]*) ([-.0-9]*) ([-.0-9]*) (|[-0-9]*) (-|[-0-9]*) ([-0-9]*) ([-0-9]*) "([^ ]*) (.*) (- |[^ ]*)" "([^"]*)" ([A-Z0-9-_]+) ([A-Za-z0-9.-]*) ([^ ]*) "([^"]*)" "([^"]*)" "([^"]*)" ([-.0-9]*) ([^ ]*) "([^"]*)" "([^"]*)" "([^ ]*)" "([^\\s]+?)" "([^\\s]+)" "([^ ]*)" "([^ ]*)" ?([^ ]*)?( .*)?',
            },
          },
          compressed: false,
          storedAsSubDirectories: false,
        },
      },
    });
  }
}
