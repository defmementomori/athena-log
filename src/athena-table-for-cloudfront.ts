import { aws_glue as glue, Stack } from 'aws-cdk-lib';
import { Construct } from 'constructs';

export interface AthenaTableForCloudFrontProps {
  /**
   * The name of the S3 bucket where the processed, Parquet-formatted CloudFront logs are stored.
   */
  readonly logBucketName: string;

  /**
   * The name of the Glue database for this table.
   */
  readonly databaseName: string;

  /**
   * The name for the Glue table.
   */
  readonly tableName: string;

  /**
   * The start date for the partition projection, in 'YYYY/MM/DD' format.
   */
  readonly projectionStartDate: string;

  /**
   * The prefix within the S3 bucket where the partitioned Parquet files are located.
   * e.g., `s3://your-bucket/my-parquet-prefix/year=...`
   * @default No prefix is used.
   */
  readonly logPrefix?: string;
}

/**
 * Creates a Glue table for querying Parquet-formatted, partitioned CloudFront logs.
 * NOTE: This construct assumes an ETL process converts raw CloudFront logs into
 * a partitioned Parquet format in S3.
 */
export class AthenaTableForCloudFront extends Construct {
  public readonly table: glue.CfnTable;

  constructor(scope: Construct, id: string, props: AthenaTableForCloudFrontProps) {
    super(scope, id);

    const { logBucketName, databaseName, tableName, projectionStartDate, logPrefix } = props;

    const stack = Stack.of(this);
    const s3LocationPrefix = logPrefix ? `${logPrefix}/` : '';
    const s3LocationTemplate = `s3://${logBucketName}/${s3LocationPrefix}year=\${year}/month=\${month}/day=\${day}/`;
    const s3BaseLocation = `s3://${logBucketName}/${s3LocationPrefix}`;

    this.table = new glue.CfnTable(this, 'Default', {
      catalogId: stack.account,
      databaseName: databaseName,
      tableInput: {
        name: tableName,
        description: 'Table for partitioned, Parquet-formatted CloudFront logs',
        tableType: 'EXTERNAL_TABLE',
        parameters: {
          'projection.enabled': 'true',
          'projection.year.type': 'integer',
          'projection.year.range': projectionStartDate.substring(0, 4) + ',2200',
          'projection.month.type': 'integer',
          'projection.month.range': '01,12',
          'projection.day.type': 'integer',
          'projection.day.range': '01,31',
          'storage.location.template': s3LocationTemplate,
        },
        partitionKeys: [
          { name: 'year', type: 'integer' },
          { name: 'month', type: 'integer' },
          { name: 'day', type: 'integer' },
        ],
        storageDescriptor: {
          columns: [
            { name: 'timestamp', type: 'timestamp' },
            { name: 'c_ip', type: 'string' },
            { name: 'time_to_first_byte', type: 'double' },
            { name: 'sc_status', type: 'int' },
            { name: 'sc_bytes', type: 'bigint' },
            { name: 'cs_method', type: 'string' },
            { name: 'cs_protocol', type: 'string' },
            { name: 'cs_host', type: 'string' },
            { name: 'cs_uri_stem', type: 'string' },
            { name: 'cs_bytes', type: 'bigint' },
            { name: 'x_edge_location', type: 'string' },
            { name: 'x_edge_request_id', type: 'string' },
            { name: 'x_host_header', type: 'string' },
            { name: 'time_taken', type: 'double' },
            { name: 'cs_protocol_version', type: 'string' },
            { name: 'c_ip_version', type: 'string' },
            { name: 'cs_user_agent', type: 'string' },
            { name: 'cs_referer', type: 'string' },
            { name: 'cs_cookie', type: 'string' },
            { name: 'cs_uri_query', type: 'string' },
            { name: 'x_edge_response_result_type', type: 'string' },
            { name: 'x_forwarded_for', type: 'string' },
            { name: 'ssl_protocol', type: 'string' },
            { name: 'ssl_cipher', type: 'string' },
            { name: 'x_edge_result_type', type: 'string' },
            { name: 'fle_encrypted_fields', type: 'int' },
            { name: 'fle_status', type: 'string' },
            { name: 'sc_content_type', type: 'string' },
            { name: 'sc_content_len', type: 'bigint' },
            { name: 'sc_range_start', type: 'bigint' },
            { name: 'sc_range_end', type: 'bigint' },
          ],
          location: s3BaseLocation,
          inputFormat: 'org.apache.hadoop.hive.ql.io.parquet.MapredParquetInputFormat',
          outputFormat: 'org.apache.hadoop.hive.ql.io.parquet.MapredParquetOutputFormat',
          serdeInfo: {
            serializationLibrary: 'org.apache.hadoop.hive.ql.io.parquet.serde.ParquetHiveSerDe',
            parameters: {
              path: 'timestamp, c_ip, time_to_first_byte, sc_status, sc_bytes, cs_method, cs_protocol, cs_host, cs_uri_stem, cs_bytes, x_edge_location, x_edge_request_id, x_host_header, time_taken, cs_protocol_version, c_ip_version, cs_user_agent, cs_referer, cs_cookie, cs_uri_query, x_edge_response_result_type, x_forwarded_for, ssl_protocol, ssl_cipher, x_edge_result_type, fle_encrypted_fields, fle_status, sc_content_type, sc_content_len, sc_range_start, sc_range_end',
            },
          },
        },
      },
    });
  }
}
