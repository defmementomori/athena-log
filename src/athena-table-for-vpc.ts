import { aws_glue as glue, Stack } from 'aws-cdk-lib';
import { Construct } from 'constructs';

/**
 * Properties for the AthenaTableForVpcFlowLog construct.
 */
export interface AthenaTableForVpcFlowLogProps {
  /**
   * The name of the S3 bucket where the VPC Flow Logs are stored.
   */
  readonly logBucketName: string;

  /**
   * The name of the Glue database where the table will be created.
   */
  readonly databaseName: string;

  /**
   * The name for the Glue table.
   */
  readonly tableName: string;

  /**
   * The prefix within the S3 bucket where logs are located, before the /AWSLogs/ part.
   * e.g., if logs are at `s3://<bucket>/your-prefix/AWSLogs/`
   * @default No prefix is used.
   */
  readonly logPrefix?: string;
}

/**
 * A CDK construct to create an AWS Glue table for querying VPC Flow Logs with Athena.
 * NOTE: This table requires you to run `MSCK REPAIR TABLE` to discover new partitions.
 */
export class AthenaTableForVpcFlowLog extends Construct {
  public readonly table: glue.CfnTable;

  constructor(scope: Construct, id: string, props: AthenaTableForVpcFlowLogProps) {
    super(scope, id);

    const { logBucketName, databaseName, tableName, logPrefix } = props;

    const stack = Stack.of(this);
    const s3LocationPrefix = logPrefix ? `${logPrefix}/` : '';
    const s3Location = `s3://${logBucketName}/${s3LocationPrefix}AWSLogs/`;

    this.table = new glue.CfnTable(this, 'Default', {
      catalogId: stack.account,
      databaseName: databaseName,
      tableInput: {
        name: tableName,
        tableType: 'EXTERNAL_TABLE',
        parameters: {
          'skip.header.line.count': '1',
        },
        partitionKeys: [
          { name: 'aws-account-id', type: 'string' },
          { name: 'aws-service', type: 'string' },
          { name: 'aws-region', type: 'string' },
          { name: 'year', type: 'string' },
          { name: 'month', type: 'string' },
          { name: 'day', type: 'string' },
          { name: 'hour', type: 'string' },
        ],
        storageDescriptor: {
          columns: [
            { name: 'version', type: 'int' },
            { name: 'account_id', type: 'string' },
            { name: 'interface_id', type: 'string' },
            { name: 'srcaddr', type: 'string' },
            { name: 'dstaddr', type: 'string' },
            { name: 'srcport', type: 'int' },
            { name: 'dstport', type: 'int' },
            { name: 'protocol', type: 'bigint' },
            { name: 'packets', type: 'bigint' },
            { name: 'bytes', type: 'bigint' },
            { name: 'start', type: 'bigint' },
            { name: 'end', type: 'bigint' },
            { name: 'action', type: 'string' },
            { name: 'log_status', type: 'string' },
            { name: 'vpc_id', type: 'string' },
            { name: 'subnet_id', type: 'string' },
            { name: 'instance_id', type: 'string' },
            { name: 'tcp_flags', type: 'int' },
            { name: 'type', type: 'string' },
            { name: 'pkt_srcaddr', type: 'string' },
            { name: 'pkt_dstaddr', type: 'string' },
            { name: 'region', type: 'string' },
            { name: 'az_id', type: 'string' },
            { name: 'sublocation_type', type: 'string' },
            { name: 'sublocation_id', type: 'string' },
            { name: 'pkt_src_aws_service', type: 'string' },
            { name: 'pkt_dst_aws_service', type: 'string' },
            { name: 'flow_direction', type: 'string' },
            { name: 'traffic_path', type: 'int' },
          ],
          location: s3Location,
          inputFormat: 'org.apache.hadoop.hive.ql.io.parquet.MapredParquetInputFormat',
          outputFormat: 'org.apache.hadoop.hive.ql.io.parquet.MapredParquetOutputFormat',
          serdeInfo: {
            serializationLibrary: 'org.apache.hadoop.hive.ql.io.parquet.serde.ParquetHiveSerDe',
          },
        },
      },
    });
  }
}
