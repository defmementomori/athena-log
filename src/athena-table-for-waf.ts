import { aws_glue as glue, Stack } from 'aws-cdk-lib';
import { Construct } from 'constructs';

export interface AthenaTableForWafProps {
  /** S3に保存されたWAFログのバケット名 */
  readonly logBucketName: string;
  /** Glueデータベース名 */
  readonly databaseName: string;
  /** 作成するGlueテーブル名 */
  readonly tableName: string;
  /**
   * WAFのスコープを指定します。
   * 'REGIONAL' または 'CLOUDFRONT'
   */
  readonly wafScope: 'REGIONAL' | 'CLOUDFRONT';
  /** クエリ対象のWeb ACL名 */
  readonly webAclName: string;
  /** バケット内のオプションのプレフィックス */
  readonly logPrefix?: string;
}

/**
 * S3に保存されたAWS WAFログをクエリするためのGlueテーブルを作成します。
 * このコンストラクトは、公式ドキュメントのdate型パーティション射影に準拠します。
 */
export class AthenaTableForWaf extends Construct {
  public readonly table: glue.CfnTable;

  constructor(scope: Construct, id: string, props: AthenaTableForWafProps) {
    super(scope, id);

    const { logBucketName, databaseName, tableName, wafScope, webAclName, logPrefix } = props;

    const stack = Stack.of(this);
    const account = stack.account;

    const logRegion = wafScope === 'CLOUDFRONT' ? 'cloudfront' : stack.region;

    const s3Prefix = logPrefix ? `${logPrefix}/` : '';
    const s3BaseLocation = `s3://${logBucketName}/${s3Prefix}AWSLogs/${account}/WAFLogs/${logRegion}/${webAclName}/`;
    const s3LocationTemplate = `${s3BaseLocation}\${log_time}`;

    this.table = new glue.CfnTable(this, 'Default', {
      catalogId: account,
      databaseName: databaseName,
      tableInput: {
        name: tableName,
        tableType: 'EXTERNAL_TABLE',
        parameters: {
          'projection.enabled': 'true',
          'projection.log_time.type': 'date',
          'projection.log_time.format': 'yyyy/MM/dd/HH/mm',
          'projection.log_time.range': '2025/01/01/00/00,NOW',
          'projection.log_time.interval': '1',
          'projection.log_time.interval.unit': 'MINUTES',
          'storage.location.template': s3LocationTemplate,
        },
        partitionKeys: [{ name: 'log_time', type: 'string' }],
        storageDescriptor: {
          columns: [
            { name: 'timestamp', type: 'bigint' },
            { name: 'formatversion', type: 'int' },
            { name: 'webaclid', type: 'string' },
            { name: 'terminatingruleid', type: 'string' },
            { name: 'terminatingruletype', type: 'string' },
            { name: 'action', type: 'string' },
            { name: 'terminatingrulematchdetails', type: 'array<struct<conditiontype:string,sensitivitylevel:string,location:string,matcheddata:array<string>>>' },
            { name: 'httpsourcename', type: 'string' },
            { name: 'httpsourceid', type: 'string' },
            {
              name: 'rulegrouplist',
              type: 'array<struct<rulegroupid:string,terminatingrule:struct<ruleid:string,action:string,rulematchdetails:array<struct<conditiontype:string,sensitivitylevel:string,location:string,matcheddata:array<string>>>>,nonterminatingmatchingrules:array<struct<ruleid:string,action:string,overriddenaction:string,rulematchdetails:array<struct<conditiontype:string,sensitivitylevel:string,location:string,matcheddata:array<string>>>,challengeresponse:struct<responsecode:string,solvetimestamp:string>,captcharesponse:struct<responsecode:string,solvetimestamp:string>>>,excludedrules:string>>',
            },
            { name: 'ratebasedrulelist', type: 'array<struct<ratebasedruleid:string,limitkey:string,maxrateallowed:int>>' },
            {
              name: 'nonterminatingmatchingrules',
              type: 'array<struct<ruleid:string,action:string,rulematchdetails:array<struct<conditiontype:string,sensitivitylevel:string,location:string,matcheddata:array<string>>>,challengeresponse:struct<responsecode:string,solvetimestamp:string>,captcharesponse:struct<responsecode:string,solvetimestamp:string>>',
            },
            { name: 'requestheadersinserted', type: 'array<struct<name:string,value:string>>' },
            { name: 'responsecodesent', type: 'string' },
            {
              name: 'httprequest',
              type: 'struct<clientip:string,country:string,headers:array<struct<name:string,value:string>>,uri:string,args:string,httpversion:string,httpmethod:string,requestid:string,fragment:string,scheme:string,host:string>',
            },
            { name: 'labels', type: 'array<struct<name:string>>' },
            { name: 'captcharesponse', type: 'struct<responsecode:string,solvetimestamp:string,failurereason:string>' },
            { name: 'challengeresponse', type: 'struct<responsecode:string,solvetimestamp:string,failurereason:string>' },
            { name: 'ja3fingerprint', type: 'string' },
            { name: 'ja4fingerprint', type: 'string' },
            { name: 'oversizefields', type: 'string' },
            { name: 'requestbodysize', type: 'int' },
            { name: 'requestbodysizeinspectedbywaf', type: 'int' },
          ],
          location: s3BaseLocation,
          serdeInfo: {
            serializationLibrary: 'org.openx.data.jsonserde.JsonSerDe',
          },
        },
      },
    });
  }
}
