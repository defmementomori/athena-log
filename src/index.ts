export * from './athena-table-for-alb';
export * from './athena-table-for-vpc';
export * from './athena-table-for-waf';
export * from './athena-table-for-cloudfront';
