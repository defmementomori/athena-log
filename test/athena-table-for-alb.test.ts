import { Stack } from 'aws-cdk-lib';
import { Template } from 'aws-cdk-lib/assertions';
import { AthenaTableForAlb } from '../src/index';

// AthenaTableForAlb Constructのテスト
test('Snapshot Test for AthenaTableForAlb', () => {
  // GIVEN (前提条件): テスト用のStackを準備
  const stack = new Stack();

  // WHEN (実行): 作成したConstructをインスタンス化
  new AthenaTableForAlb(stack, 'MyTestTable', {
    logBucketName: 'test-log-bucket',
    databaseName: 'test_database',
    tableName: 'test_table',
    projectionStartDate: '2025-08-01',
  });

  // THEN (検証): 生成されたCloudFormationテンプレートが期待通りか確認
  const template = Template.fromStack(stack);

  // 生成されたテンプレートが、保存されたスナップショットと一致するか検証
  expect(template.toJSON()).toMatchSnapshot();
});
