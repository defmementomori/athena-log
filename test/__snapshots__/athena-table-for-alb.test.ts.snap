// Jest Snapshot v1, https://jestjs.io/docs/snapshot-testing

exports[`Snapshot Test for AthenaTableForAlb 1`] = `
{
  "Parameters": {
    "BootstrapVersion": {
      "Default": "/cdk-bootstrap/hnb659fds/version",
      "Description": "Version of the CDK Bootstrap resources in this environment, automatically retrieved from SSM Parameter Store. [cdk:skip]",
      "Type": "AWS::SSM::Parameter::Value<String>",
    },
  },
  "Resources": {
    "MyTestTable": {
      "Properties": {
        "CatalogId": {
          "Ref": "AWS::AccountId",
        },
        "DatabaseName": "test_database",
        "TableInput": {
          "Description": "Table for querying ALB access logs",
          "Name": "test_table",
          "Parameters": {
            "EXTERNAL": "TRUE",
            "has_encrypted_data": "false",
            "projection.day.format": "yyyy/MM/dd",
            "projection.day.interval": "1",
            "projection.day.interval.unit": "DAYS",
            "projection.day.range": "2025-08-01,2200",
            "projection.day.type": "date",
            "projection.enabled": "true",
            "storage.location.template": {
              "Fn::Join": [
                "",
                [
                  "s3://test-log-bucket/AWSLogs/",
                  {
                    "Ref": "AWS::AccountId",
                  },
                  "/elasticloadbalancing/",
                  {
                    "Ref": "AWS::Region",
                  },
                  "/\${day}/",
                ],
              ],
            },
          },
          "PartitionKeys": [
            {
              "Name": "day",
              "Type": "string",
            },
          ],
          "StorageDescriptor": {
            "Columns": [
              {
                "Name": "type",
                "Type": "string",
              },
              {
                "Name": "time",
                "Type": "string",
              },
              {
                "Name": "elb",
                "Type": "string",
              },
              {
                "Name": "client_ip",
                "Type": "string",
              },
              {
                "Name": "client_port",
                "Type": "int",
              },
              {
                "Name": "target_ip",
                "Type": "string",
              },
              {
                "Name": "target_port",
                "Type": "int",
              },
              {
                "Name": "request_processing_time",
                "Type": "double",
              },
              {
                "Name": "target_processing_time",
                "Type": "double",
              },
              {
                "Name": "response_processing_time",
                "Type": "double",
              },
              {
                "Name": "elb_status_code",
                "Type": "int",
              },
              {
                "Name": "target_status_code",
                "Type": "string",
              },
              {
                "Name": "received_bytes",
                "Type": "bigint",
              },
              {
                "Name": "sent_bytes",
                "Type": "bigint",
              },
              {
                "Name": "request_verb",
                "Type": "string",
              },
              {
                "Name": "request_url",
                "Type": "string",
              },
              {
                "Name": "request_proto",
                "Type": "string",
              },
              {
                "Name": "user_agent",
                "Type": "string",
              },
              {
                "Name": "ssl_cipher",
                "Type": "string",
              },
              {
                "Name": "ssl_protocol",
                "Type": "string",
              },
              {
                "Name": "target_group_arn",
                "Type": "string",
              },
              {
                "Name": "trace_id",
                "Type": "string",
              },
              {
                "Name": "domain_name",
                "Type": "string",
              },
              {
                "Name": "chosen_cert_arn",
                "Type": "string",
              },
              {
                "Name": "matched_rule_priority",
                "Type": "string",
              },
              {
                "Name": "request_creation_time",
                "Type": "string",
              },
              {
                "Name": "actions_executed",
                "Type": "string",
              },
              {
                "Name": "redirect_url",
                "Type": "string",
              },
              {
                "Name": "lambda_error_reason",
                "Type": "string",
              },
              {
                "Name": "target_port_list",
                "Type": "string",
              },
              {
                "Name": "target_status_code_list",
                "Type": "string",
              },
              {
                "Name": "classification",
                "Type": "string",
              },
              {
                "Name": "classification_reason",
                "Type": "string",
              },
              {
                "Name": "conn_trace_id",
                "Type": "string",
              },
            ],
            "Compressed": false,
            "InputFormat": "org.apache.hadoop.mapred.TextInputFormat",
            "Location": {
              "Fn::Join": [
                "",
                [
                  "s3://test-log-bucket/AWSLogs/",
                  {
                    "Ref": "AWS::AccountId",
                  },
                  "/elasticloadbalancing/",
                  {
                    "Ref": "AWS::Region",
                  },
                  "/",
                ],
              ],
            },
            "OutputFormat": "org.apache.hadoop.hive.ql.io.HiveIgnoreKeyTextOutputFormat",
            "SerdeInfo": {
              "Parameters": {
                "input.regex": "([^ ]*) ([^ ]*) ([^ ]*) ([^ ]*):([0-9]*) ([^ ]*)[:-]([0-9]*) ([-.0-9]*) ([-.0-9]*) ([-.0-9]*) (|[-0-9]*) (-|[-0-9]*) ([-0-9]*) ([-0-9]*) "([^ ]*) (.*) (- |[^ ]*)" "([^"]*)" ([A-Z0-9-_]+) ([A-Za-z0-9.-]*) ([^ ]*) "([^"]*)" "([^"]*)" "([^"]*)" ([-.0-9]*) ([^ ]*) "([^"]*)" "([^"]*)" "([^ ]*)" "([^\\s]+?)" "([^\\s]+)" "([^ ]*)" "([^ ]*)" ?([^ ]*)?( .*)?",
                "serialization.format": "1",
              },
              "SerializationLibrary": "org.apache.hadoop.hive.serde2.RegexSerDe",
            },
            "StoredAsSubDirectories": false,
          },
          "TableType": "EXTERNAL_TABLE",
        },
      },
      "Type": "AWS::Glue::Table",
    },
  },
  "Rules": {
    "CheckBootstrapVersion": {
      "Assertions": [
        {
          "Assert": {
            "Fn::Not": [
              {
                "Fn::Contains": [
                  [
                    "1",
                    "2",
                    "3",
                    "4",
                    "5",
                  ],
                  {
                    "Ref": "BootstrapVersion",
                  },
                ],
              },
            ],
          },
          "AssertDescription": "CDK bootstrap stack version 6 required. Please run 'cdk bootstrap' with a recent version of the CDK CLI.",
        },
      ],
    },
  },
}
`;
